File: Devoir2/code/Canal.java
package code;
import java.util.Random;


/**
 * Canal.java
 * Simulation d'un canal de transmission non fiable.
 * Chaque trame ou ACK peut être perdue, corrompue ou retardée.
 * Paramètres configurables : probErreur, probPerte, delaiMax.
 */

public class Canal {

    // Paramètres du canal
    private double probErreur = 0.05;
    private double probPerte = 0.10;
    private int delaiMax = 200; // en millisecondes
    private Random rand = new Random();

    public Canal(double probErreur, double probPerte, int delaiMax) {
        this.probErreur = probErreur;
        this.probPerte = probPerte;
        this.delaiMax = delaiMax;
    }

    public Canal() {
        // Constructeur par défaut avec paramètres par défaut
        this(0.05, 0.10, 200);
    }

    // Simule la transmission d'une trame du côté émetteur vers le récepteur
    public void transmettre (Trame trame) {
        double tirage = rand.nextDouble();

        // 1 - Simule la perte de trame
        if (tirage < probPerte) {
            System.out.println("Canal : Trame perdue : " + trame);
            return;
        }

        // 2 - Simule la corruption de trame
        if (tirage < probPerte + probErreur) {
            corrompre(trame);
            System.out.println("Canal : Trame corrompue : " + trame);
        }

        // 3 - Simule le délai de transmission
        int delai = rand.nextInt(delaiMax + 1);
        try {
            // Pause pour simuler le délai
            Thread.sleep(delai);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        // Livraison de la trame
        System.out.println("Canal: Trame transmise après " + delai + " ms : " + trame);
    }

    // Méthode pour corrompre une trame en modifiant un caractère aléatoire
    private void corrompre(Trame trame) {
        String data = trame.getDonnees();
        if (data.isEmpty()) return;
        char[] chars = data.toCharArray();
        int pos = rand.nextInt(chars.length);
        chars[pos] = (char) (chars[pos] + 1);
        trame.setDonnees(new String(chars));
    }

    // Test
    public static void main(String[] args) {
        // Création d'un canal
        Canal canal = new Canal();
        // Envoi de 100 trames pour test
        for (int i = 0; i < 100; i++) {
            Trame trame = new Trame(i, "Trame " + i, false);
            canal.transmettre(trame);
        }
        System.out.println("Simulation de canal terminée.");
    }



}



    


    

File: Devoir2/code/Canal.py
import random
import threading

class Canal:
    """
    Canal de transmission générique simulant un lien non fiable.
    Les paramètres probErreur, probPerte et delaiMax sont configurables
    pour reproduire différents scénarios.
    """

    def __init__(self, probErreur=0.0, probPerte=0.0, delaiMax=0):
        """
        Initialise le canal.
        probErreur : probabilité de corruption d'une trame (0.0 à 1.0)
        probPerte  : probabilité de perte d'une trame
        delaiMax   : délai maximal (en ms)
        """
        self.probErreur = probErreur
        self.probPerte = probPerte
        self.delaiMax = delaiMax
        self._fifo = []  # File FIFO pour garder l’ordre
        print(f"[Canal] Initialisé → erreur={probErreur}, perte={probPerte}, delaiMax={delaiMax} ms")

    def transmettre(self, trame, callback_reception):
        """
        Simule la transmission d'une trame ou d'un ACK via le canal.
        Le callback est appelé quand la trame arrive au récepteur.
        """
        tirage = random.random()

        #  Trame perdue
        if tirage < self.probPerte:
            print(f"[Canal]  Trame perdue : {trame}")
            return

        #  Trame corrompue
        if tirage < self.probPerte + self.probErreur:
            trame = self._corrompre(trame)
            print(f"[Canal] Trame corrompue : {trame}")

        # Délai simulé
        delai = random.randint(0, self.delaiMax) if self.delaiMax > 0 else 0
        self._fifo.append(trame)
        print(f"[Canal]  Transmission (+{delai} ms) : {trame}")
        threading.Timer(delai / 1000, self._livrer, args=(callback_reception,)).start()

    def _livrer(self, callback_reception):
        """Livre la trame la plus ancienne (ordre préservé)."""
        if not self._fifo:
            return
        trame = self._fifo.pop(0)
        print(f"[Canal]  Trame livrée : {trame}")
        callback_reception(trame)

    def _corrompre(self, trame):
        """Altère un caractère dans le champ données (si présent)."""
        if hasattr(trame, "donnees") and trame.donnees:
            pos = random.randint(0, len(trame.donnees) - 1)
            c = trame.donnees[pos]
            trame.donnees = trame.donnees[:pos] + chr((ord(c) + 1) % 256) + trame.donnees[pos+1:]
        return trame
